### PR TITLE
remove repr mixin

### PR DESCRIPTION
**Pull request by KnathanM**
Removes `ReprMixin` . See #649 for discussion.

### CODE REVIEW

File: chemprop/nn/loss.py [hunk 1]
```diff
@@ -23,13 +22,12 @@
...
     "BinaryDirichletLoss",
     "MulticlassDirichletLoss",
-    "_ThresholdMixin",
     "SIDLoss",
     "WassersteinLoss",
 ]
 
 
-class LossFunction(ABC, ReprMixin):
+class LossFunction(ABC, nn.Module):
```
Comment (davidegraff) [inline chemprop/nn/loss.py:30]:
```suggestion
class LossFunction(nn.Module):
```
`Module`s are `ABC`s

File: chemprop/nn/loss.py [hunk 2]
```diff
@@ -75,6 +73,13 @@ def __call__(
     def forward(self, preds, targets, mask, w_s, w_t, lt_mask, gt_mask) -> Tensor:
         """Calculate a tensor of shape `b x t` containing the unreduced loss values."""
 
+    def extra_repr(self) -> str:
+        sig_params = inspect.signature(self.__class__).parameters
+        obj_params = self.__dict__.items()
+        items = [(k, v) for k, v in obj_params if (k in sig_params and v != sig_params[k].default)]
+
+        return ", ".join(f"{k}={repr(v)}" for k, v in items)
```
Comment (davidegraff) [inline chemprop/nn/loss.py:81]: Feel like it's not worth it to set up a generic implementation of the function. Rather, each class should define its own.

File: chemprop/nn/loss.py [hunk 3]
```diff
@@ -293,16 +294,12 @@ def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor
...
-    def get_params(self) -> list[tuple[str, float]]:
-        return [("threshold", self.threshold)]
-
-
 @LossFunctionRegistry.register("sid")
-class SIDLoss(LossFunction, _ThresholdMixin):
+class SIDLoss(LossFunction):
+    def __init__(self, threshold: float | None = None):
+        super().__init__()
+        self.threshold = threshold
```
Comment (davidegraff) [inline chemprop/nn/loss.py:301]:
```suggestion

        self.threshold = threshold
```

File: chemprop/nn/loss.py [hunk 4]
```diff
@@ -316,7 +313,11 @@ def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor
 
 
 @LossFunctionRegistry.register(["earthmovers", "wasserstein"])
-class WassersteinLoss(LossFunction, _ThresholdMixin):
+class WassersteinLoss(LossFunction):
+    def __init__(self, threshold: float | None = None):
+        super().__init__()
+        self.threshold = threshold
```
Comment (davidegraff) [inline chemprop/nn/loss.py:319]:
```suggestion
        super().__init__()
        
        self.threshold = threshold
```

File: chemprop/nn/loss.py [hunk 5]
```diff
@@ -23,13 +22,12 @@
...
     "MulticlassDirichletLoss",
-    "_ThresholdMixin",
     "SIDLoss",
     "WassersteinLoss",
 ]
 
 
-class LossFunction(ABC, ReprMixin):
+class LossFunction(ABC, nn.Module):
     def __call__(
```
Comment (davidegraff) [inline chemprop/nn/loss.py:31]: we'll probably have to rename this to `forward()` and the original `forward()` method to `_forward()` or `_calc_unreduced_loss()` to adhere to the `nn.Module` interface.

File: chemprop/nn/loss.py (hunk 2 above)
Comment (KnathanM) [reply chemprop/nn/loss.py:81]: I'm not sure how to write a class specific implementation. We will still need to call `inspect.signature` to get the defaults (so that we don't print the value if it is the default). We will also still need to use `self.__dict__.items()` to get the current values. 
I think the only benefit of a class specific versus general version is we wouldn't need to check `k in sig_params` because we know what the keys should be. 
Given the current generic implementation works and reduces copying code, I prefer not switching to class implementations.

Comment (davidegraff) [reply chemprop/nn/loss.py:81]: Ex:

```python
class SIDLoss(nn.Module):
    def extra_repr(self):
        if self.threshold is None:
            return f"threshold={self.threshold}" 

        return f"threshold={self.threshold:0.2f}" 
...
loss = SIDLoss(threshold=0.5)
print(loss)
# SIDLoss(threshold=0.5)
```

Finding some generic way to implmenet `extra_repr` is probably more trouble than it's worth is my point. Using the `inspect` module is a fairly brittle to dynamically generate the string, and I don't think it's that bad to actually just hard-code the behavior in subclasses that would benefit from it.

**Comment by davidegraff** [comment]
You're getting errors now because `Metric` inherits from `LossFunction`, which is now a `Module` subclass. So you'll also have to rename the appropriate methods (`__call__()` and `forward()`) in the `Metric` class.

File: chemprop/nn/loss.py [hunk 6]
```diff
@@ -152,19 +150,25 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
...
-    def get_params(self) -> list[tuple[str, float]]:
-        return [("v_kl", self.v_kl), ("eps", self.eps)]
+    def extra_repr(self) -> str:
+        if self.v_kl != 0.2 and self.eps != 1e-8:
+            return f"v_kl={self.v_kl}, eps={self.eps}"
+        elif self.v_kl != 0.2:
+            return f"v_kl={self.v_kl}"
+        elif self.eps != 1e-8:
+            return f"eps={self.eps}"
+        return ""
```
Comment (davidegraff) [inline chemprop/nn/loss.py:160]:
```suggestion
       return f"v_kl={self.v_kl}, eps={self.eps}"
```
Omitting default values is sklearn style. torch style `repr()` includes them:
```python
>>> nn.Dropout()
Dropout(p=0.5)
```

File: chemprop/nn/loss.py [hunk 7]
```diff
@@ -270,40 +275,39 @@ def forward(self, preds, targets, *args) -> Tensor:
 
         return (L_mse + self.v_kl * L_kl).mean(-1)
 
-    def get_params(self) -> list[tuple[str, float]]:
-        return [("v_kl", self.v_kl)]
+    def extra_repr(self) -> str:
+        if self.v_kl != 0.2:
+            return f"v_kl={self.v_kl}"
+        return ""
```
Comment (davidegraff) [inline chemprop/nn/loss.py:281]:
```suggestion
        return f"v_kl={self.v_kl}"
```

File: chemprop/nn/loss.py [hunk 8]
```diff
@@ -314,13 +318,28 @@ def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor
 
         return (preds_norm / targets).log() * preds_norm + (targets / preds_norm).log() * targets
 
+    def extra_repr(self) -> str:
+        if self.threshold is not None:
+            return f"threshold={self.threshold}"
+        return ""
```
Comment (davidegraff) [inline chemprop/nn/loss.py:324]:
```suggestion
        return f"threshold={self.threshold}"
```

File: chemprop/nn/loss.py [hunk 9]
```diff
@@ -314,13 +318,28 @@ def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor
...
             preds = preds.clamp(min=self.threshold)
 
         preds_norm = preds / (preds * mask).sum(1, keepdim=True)
 
         return (targets.cumsum(1) - preds_norm.cumsum(1)).abs()
+
+    def extra_repr(self) -> str:
+        if self.threshold is not None:
+            return f"threshold={self.threshold}"
+        return ""
```
Comment (davidegraff) [inline chemprop/nn/loss.py:345]:
```suggestion
        return f"threshold={self.threshold}"

```

**Review by davidegraff — APPROVED** [review]
looks like tests are passing now

File: chemprop/nn/loss.py (hunk 6 above)
Comment (KnathanM) [reply chemprop/nn/loss.py]: Good job torch, I preferred not omitting defaults, so will commit these.

**Review by shihchengli — APPROVED** [review]
LGTM
